### PR TITLE
async t()

### PR DESCRIPTION
**Pull request by JanTrotnow**

### CODE REVIEW

File: src/components/state-machine/base-state.ts [hunk 1]
```diff
@@ -70,8 +70,8 @@ export abstract class BaseState implements State.Required, Voiceable, TranslateH
   }
 
   /** Prompts with current unhandled message */
-  public unhandledGenericIntent(machine: Transitionable, originalIntentMethod: string, ...args: any[]): any {
-    this.responseFactory.createVoiceResponse().prompt(this.translateHelper.t());
+  async unhandledGenericIntent(machine: Transitionable, originalIntentMethod: string, ...args: any[]): Promise<any> {
```
Comment (antoniusostermann) [inline src/components/state-machine/base-state.ts]: Re-add `public` everywhere, see above.

File: src/components/state-machine/base-state.ts [hunk 2]
```diff
@@ -84,27 +84,27 @@ export abstract class BaseState implements State.Required, Voiceable, TranslateH
...
    * If you pass an absolute key (without "." at beginning), this method will look at given absolute key.
    * @param locals Variables to use in reponse
    */
-  public t(key?: string, locals?: { [name: string]: string | number | object }): string;
+  t(key?: string, locals?: {[name: string]: string | number | object}): Promise<string>;
 
-  public t(...args: any[]) {
-    return (this.translateHelper as any).t(...args);
+  async t(...args: any[]) {
+    return await (this.translateHelper as any).t(...args);
```
Comment (antoniusostermann) [inline src/components/state-machine/base-state.ts]: Returning the `await` is unnecessary. 

https://github.com/eslint/eslint/blob/master/docs/rules/no-return-await.md

If not already present in linter, we should add this rule in both tslints.

File: src/components/unifier/public-interfaces.ts [hunk 3]
```diff
@@ -48,14 +48,14 @@ export interface Voiceable {
...
-  endSessionWith(text: string): void;
+  endSessionWith(text: string): void | Promise<void>;
 
   /**
    * Sends voice message but does not end session, so the user is able to respond
    * @param {string} text Text to say to user
    * @param {string[]} [reprompts] If the user does not answer in a given time, these reprompt messages will be used.
    */
-  prompt(text: string, ...reprompts: string[]): void;
+  prompt(text: string | Promise<string>, ...reprompts: Array<string | Promise<string>>): void | Promise<void>;
```
Comment (antoniusostermann) [inline src/components/unifier/public-interfaces.ts]: We should change the return values of this method. To make it easy for a developer, we should *accept* `string` and `Promise<string>` (as you've done it), but *return* always `Promise<void>`.

File: src/components/unifier/public-interfaces.ts [hunk 4]
```diff
@@ -48,14 +48,14 @@ export interface Voiceable {
    * Sends voice message and ends session
    * @param {string} text Text to say to user
    */
-  endSessionWith(text: string): void;
+  endSessionWith(text: string): void | Promise<void>;
```
Comment (antoniusostermann) [inline src/components/unifier/public-interfaces.ts]: You should also accept `Promise<string>`.

We should change the return values of this method. To make it easy for a developer, we should *accept* `string` and `Promise<string>` (as you've done it), but *return* always `Promise<void>`.

File: src/components/unifier/responses/simple-voice-response.ts [hunk 5]
```diff
@@ -16,11 +16,82 @@ export class SimpleVoiceResponse extends BaseResponse implements Voiceable {
     this.handler.sendResponse();
   }
 
-  public prompt(text: string, ...reprompts: string[]) {
+  prompt(text: string | Promise<string>, ...reprompts: Array<string | Promise<string>>): void | Promise<void> {
     this.handler.endSession = false;
-    this.handler.voiceMessage = this.prepareText(text);
-    this.attachRepromptsIfAny(reprompts);
-    this.handler.sendResponse();
+    let promiseA: Promise<any> | undefined;
```
Comment (antoniusostermann) [inline src/components/unifier/responses/simple-voice-response.ts]: This implementation is *way too complicated*. Reconsider your implementation using `Promise.resolve`. You should only need a few lines for this (as it was before).

File: src/components/unifier/responses/voice-response.ts [hunk 6]
```diff
@@ -10,17 +10,37 @@ export class VoiceResponse implements Voiceable {
...
 
-  public endSessionWith(text: string) {
-    this.delegatorBasedOnInput(text).endSessionWith(text);
+  endSessionWith(text: string) {
+    if(typeof text === "string") {
+      return (this.delegatorBasedOnInput(text) as Voiceable).endSessionWith(text);
+    }
+    return (this.delegatorBasedOnInput(text) as Promise<Voiceable>).then((voiceable) => {
+      return voiceable.endSessionWith(text);
+    });
```
Comment (antoniusostermann) [inline src/components/unifier/responses/voice-response.ts]: Use `Promise.resolve`!

File: src/components/unifier/responses/voice-response.ts [hunk 7]
```diff
@@ -10,17 +10,37 @@ export class VoiceResponse implements Voiceable {
...
 
-  public prompt(text: string, ...reprompts: string[]) {
-    this.delegatorBasedOnInput(text).prompt(text, ...reprompts);
+  prompt(text: string | Promise<string>, ...reprompts: Array<string | Promise<string>>) {
+    if (typeof text === "string") {
+      return (this.delegatorBasedOnInput(text) as Voiceable).prompt(text, ...reprompts);
+    }
+    return (this.delegatorBasedOnInput(text) as Promise<Voiceable>).then(voiceable => {
+      return voiceable.prompt(text, ...reprompts);
+    });
```
Comment (antoniusostermann) [inline src/components/unifier/responses/voice-response.ts]: Use `Promise.resolve`!

File: src/components/unifier/responses/voice-response.ts [hunk 8]
```diff
@@ -10,17 +10,37 @@ export class VoiceResponse implements Voiceable {
...
   }
 
-  private delegatorBasedOnInput(text: string) {
+  private delegatorBasedOnInput(text: string | Promise<string>) {
     if (typeof text === "undefined" || text === null) text = "";
 
+    if (typeof text !== "string") {
+      return text.then(value => {
+        if (value.includes("</") || value.includes("/>")) {
```
Comment (antoniusostermann) [inline src/components/unifier/responses/voice-response.ts]: favor `async/await`

File: src/components/unifier/responses/voice-response.ts [hunk 9]
```diff
@@ -10,17 +10,37 @@ export class VoiceResponse implements Voiceable {
...
+    if (typeof text === "string") {
+      return (this.delegatorBasedOnInput(text) as Voiceable).prompt(text, ...reprompts);
+    }
+    return (this.delegatorBasedOnInput(text) as Promise<Voiceable>).then(voiceable => {
+      return voiceable.prompt(text, ...reprompts);
+    });
   }
 
-  private delegatorBasedOnInput(text: string) {
+  private delegatorBasedOnInput(text: string | Promise<string>) {
```
Comment (antoniusostermann) [inline src/components/unifier/responses/voice-response.ts]: This is an internal method, accept `Promise<string>` only.

File: src/components/i18n/interpolation-resolver.ts [hunk 10]
```diff
@@ -24,28 +24,29 @@ export class InterpolationResolver implements InterpolationResolverInterface {
...
   public async resolveMissingInterpolations(translatedValue: string, translateHelper: TranslateHelper): Promise<string> {
-    while (translatedValue.includes("*~~") && translatedValue.includes("~~*")) {
-      const interpolation = translatedValue.split("*~~")[1].split("~~*")[0];
-      let interpolationValue: string | undefined;
+    let text = translatedValue;
 
-      for (let missingInterpolationExtension of this.missingInterpolationExtensions) {
-        interpolationValue = await missingInterpolationExtension.execute(interpolation, translateHelper);
+    while (text.includes("*~~") && text.includes("~~*")) {
+      const interpolation = text.split("*~~")[1].split("~~*")[0];
```
[the remark this answers is not in the document]
Comment (JanTrotnow) [reply src/components/i18n/interpolation-resolver.ts]: I did it this way to be able to handle possible missing interpolations that are returned by an extension.

e.g. the extension resolving {{thing}} in "i like {{thing}}" might return a text containing an interpolation like "{{flavour}} ice cream" aswell.

This way im able to resolve all occurring interpolations iteratively.

**Pull request merged**